cdap-api-common/src/main/java/io/cdap/cdap/api/common/Constants.java

```diff
@@ -25,7 +25,9 @@
  * Constants used by different systems are all defined here.
  */
 public class Constants {
-  public static final List<Integer> RETRYABLE_HTTP_CODES = Collections.unmodifiableList(
-    Arrays.asList(HttpURLConnection.HTTP_BAD_GATEWAY, HttpURLConnection.HTTP_UNAVAILABLE,
-                  HttpURLConnection.HTTP_GATEWAY_TIMEOUT));
+  public static final List<Integer> HTTP_SERVER_ERROR_CODES = Collections.unmodifiableList(
+    Arrays.asList(HttpURLConnection.HTTP_BAD_GATEWAY,
+                  HttpURLConnection.HTTP_GATEWAY_TIMEOUT,
+                  HttpURLConnection.HTTP_INTERNAL_ERROR,
```

Comment:
I don't think a 500 should be treated like these others. The other ones look like they're all about unavailability, whereas internal error can mean a lot of different things. For example, a 500 can come from a bug in the code like an unexpected null pointer exception, where the request will always fail on retry.

Reply:
I believe in this case we are handling IVP problems, but overall 500 can mean an unhealthy instance (e.g. OOM) and can succeed on retry

Reply:
It's not much different from GATEWAY_TIMEOUT. GATEWAY_TIMEOUT means server did not answer in a timely fashion, while INTERNAL_ERROR means server answered with a failure. Both may indicate a retryable error (specific instance has problems) or non-retryable error (e.g. infinite loop for GATEWAY_TIMEOUT or NPE for 500). Unless we can differenciate retryable vs non-retryable better, IMHO it's good to retry more and not less